modules/acm/src/main/java/edu/gemini/epics/acm/CaTaskControlImpl.java

```diff
@@ -0,0 +1,477 @@
+        }
+
+        @Override
+        public State onTimeout() {
+            failCommand(cm, new TimeoutException());
+            return IdleState;
+        }
+    }
+
+    private final class WaitCompletion implements State {
+        private final CaCommandMonitorImpl cm;
+
+        WaitCompletion(CaCommandMonitorImpl cm) {
+            this.cm = cm;
+        }
+
+        @Override
+        public String signature() {
+            return "WaitCompletion";
+        }
+
+        @Override
+        public State onValChange(Integer val) {
+            if(val == 0) {
+                return this;
+            } else {
+                failCommandWithErrorMessage(cm);
+                return IdleState;
+            }
+        }
+
+        @Override
+        public State onStateChange(TaskControlState st) {
+            if(st == TaskControlState.ERR) {
+                failCommandWithErrorMessage(cm);
+                return IdleState;
+            } else if(st == TaskControlState.IDLE) {
+                succedCommand(cm);
+                return IdleState;
+            } else {
+                return this;
+            }
+        }
+
+        @Override
+        public State onTimeout() {
+            failCommand(cm, new TimeoutException());
+            return IdleState;
+        }
+    }
+
+    private synchronized void onValChange(final Integer val) {
+        if (val != null) {
+            State oldState = currentState;
+            currentState = currentState.onValChange(val);
+            if (currentState.equals(IdleState) && timeoutFuture != null) {
+                timeoutFuture.cancel(true);
+                timeoutFuture = null;
+            }
+            if(trace) LOG.debug("onValChange(" + val + "): " + oldState.signature() + " -> " + currentState.signature());
+        }
+    }
+
+    private synchronized void onStateChange(final TaskControlState val) {
+        if (val != null) {
+            State oldState = currentState;
+            currentState = currentState.onStateChange(val);
+            if (currentState.equals(IdleState) && timeoutFuture != null) {
+                timeoutFuture.cancel(true);
+                timeoutFuture = null;
+            }
+            if(trace) LOG.debug("onStateChange(" + val + "): " + oldState.signature() + " -> " + currentState.signature());
+        }
+    }
+
+    private synchronized void onTimeout() {
+        timeoutFuture = null;
+        State oldState = currentState;
+        currentState = currentState.onTimeout();
+        if(trace) LOG.debug("onTimeout: " + oldState.signature() + " -> " + currentState.signature());
+    }
+
+    private void succedCommand(final CaCommandMonitorImpl cm) {
+        executor.execute(new Runnable() {
+            @Override
+            public void run() {
+                cm.completeSuccess();
+            }
+        });
+    }
+
+    private void failCommand(final CaCommandMonitorImpl cm, final Exception ex) {
+        executor.execute(new Runnable() {
```

Comment:
Implementing `Runnable` like this is also overkill now. Since `Runnable` is functional interface, you can write this as:

```
executor.execute(() -> { cm.completeFailure(ex); });
```